pkg/operators/kubenetworkmanager/kubenetworkmanager.go

```diff
@@ -0,0 +1,295 @@
+		cache.Close()
+	}
+	cache.useCount--
+}
+
+func (cache *k8sInventoryCache) update() {
+	pods, err := cache.clientset.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		log.Errorf("listing pods: %v", err)
+		return
+	}
+	cache.pods.Store(pods)
+
+	svcs, err := cache.clientset.CoreV1().Services("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		log.Errorf("listing services: %s", err)
+		return
+	}
+	cache.svcs.Store(svcs)
+}
+
+func (cache *k8sInventoryCache) GetPods() *v1.PodList {
+	return cache.pods.Load()
+}
+
+func (cache *k8sInventoryCache) GetSvcs() *v1.ServiceList {
+	return cache.svcs.Load()
+}
+
+type KubeNetworkManager struct {
+	k8sInventory *k8sInventoryCache
+}
+
+func (k *KubeNetworkManager) Name() string {
+	return "KubeNetworkManager"
+}
+
+func (k *KubeNetworkManager) Description() string {
+	return "KubeNetworkManager resolves IP addresses to pod- and service-names"
+}
+
+func (k *KubeNetworkManager) GlobalParamDescs() params.ParamDescs {
+	return nil
+}
+
+func (k *KubeNetworkManager) ParamDescs() params.ParamDescs {
+	return nil
+}
+
+func (k *KubeNetworkManager) Dependencies() []string {
+	return []string{"KubeManager"}
+}
+
+func (k *KubeNetworkManager) CanOperateOn(gadget gadgets.GadgetDesc) bool {
+	km := kubemanager.KubeManager{}
+	if !km.CanOperateOn(gadget) {
+		return false
+	}
+
+	_, hasAdditionalInf := gadget.EventPrototype().(AdditionalPodNetworkInformation)
+	return hasAdditionalInf
+}
+
+func (k *KubeNetworkManager) Init(params *params.Params) error {
+	k8sInventory, err := newCache(1 * time.Second)
+	if err != nil {
+		return fmt.Errorf("creating new k8sInventoryCache: %w", err)
+	}
+	k.k8sInventory = k8sInventory
+	return nil
+}
+
+func (k *KubeNetworkManager) Close() error {
+	k.k8sInventory.Close()
+	return nil
+}
+
+func (k *KubeNetworkManager) Instantiate(gadgetCtx operators.GadgetContext, gadgetInstance any, params *params.Params) (operators.OperatorInstance, error) {
+	traceInstance := &KubeNetworkManagerInstance{
+		gadgetCtx: gadgetCtx,
+		manager:   k,
+		gadget:    gadgetInstance,
+	}
+
+	return traceInstance, nil
+}
+
+type KubeNetworkManagerInstance struct {
+	gadgetCtx operators.GadgetContext
+	manager   *KubeNetworkManager
+	gadget    any
+}
+
+func (m *KubeNetworkManagerInstance) Name() string {
+	return "KubeNetworkManagerInstance"
+}
+
+func (m *KubeNetworkManagerInstance) PreGadgetRun() error {
+	m.manager.k8sInventory.Start()
+	return nil
+}
+
+func (m *KubeNetworkManagerInstance) PostGadgetRun() error {
+	m.manager.k8sInventory.Stop()
+	return nil
+}
+
+func (m *KubeNetworkManagerInstance) enrich(ev any) {
+	additionalInfo, _ := ev.(AdditionalPodNetworkInformation)
+	containerInfo, _ := ev.(operators.ContainerInfoGetters)
+
+	additionalInfo.SetRemoteKind(types.RemoteKindOther)
+
+	pods := m.manager.k8sInventory.GetPods()
+	foundLocal := false
+	foundRemote := false
+	remoteIP := additionalInfo.GetRemoteIP()
+
+	for i, pod := range pods.Items {
```

Comment:
I don't know whether we should handle it in this PR, but:

Right now, we're iterating over all entries of the raw `pods.Items` for each event (which is hot code) - that might be a pretty long list. IMHO, it would be much more efficient to prepare lookup maps in `cache.update()`:

* a map with key of type `struct { Namespace string, Name string }` pointing to the pod
* a map with just a string as key (namely the `PodIP` or `ClusterIP`) pointing to either a pod or service

That would lead to one or two map lookups per event instead of iterating and checking possibly hundreds of pods/services.

Reply:
It is possible to do that. But before i implemented the caching mechanism the gadget was really slow and kind of "laggy".
I think iterating over thousands of pods is way faster than an API call

Reply:
I was referring to the last commit with the caching mechanism. There the API call is being done in the Background in another goroutine. Having synchronous API calls in the enricher/operator slows down the whole process by a lot.
Nonetheless I will do some quick benchmarks

Reply:
I tested with 1 and with 100 iterations for the following path:
```golang
_, err := m.manager.k8sInventory.clientset.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
if err != nil {
	log.Errorf("listing pods: %v", err)
	return
}
// -------- against ---------
// 10000 to simulate many pods
for i := 0; i < 10000; i++ {
	for _, pod := range podss.Items {
		if pod.Namespace == "default" && pod.Name == "foobar" {
			continue
  		}
  	}
  }
```
### Only 1 iteration
```
time="2023-03-17T09:38:28Z" level=error msg="Elapsed mean time for 1 API call: 3.046218ms"
time="2023-03-17T09:38:28Z" level=error msg="Elapsed mean time for iterating over 90000 pods: 1.280459ms"
```
### With 100 iterations
```
time="2023-03-17T10:33:42Z" level=error msg="Elapsed mean time for 1 API call: 394.446466ms"
time="2023-03-17T10:33:43Z" level=error msg="Elapsed mean time for iterating over 90000 pods: 1.19797ms"
```
### Conclusion (for me)
The API is slower as seen in **Only 1 iteration**. With higher pressure on the API the times are getting even worse.
Of course only when this benchmark measures the right thing. And its not that scientific either

Reply:
> The only advantage of the API here would be to reduce the memory footprint as it would returns one pod but I do not think this is a problem here?

`sizeof` gave me 1048 Bytes for the `pod struct`, which is quite large. Especially in the loops like @flyth mentioned because of copying. Measurements still say that it is faster.
Regarding general memory footprint: Lets simplify that a `pod struct` is 1024 Bytes = 1KB.
- 1 Pod = 1KB
- 1,000 Pods = 1MB
- 1,000,000 Pods = 1GB

I think the footprint is acceptable?

> And the caching mechanism you wrote should do the trick so we avoid having API calls on the hot path!

Correct, everything is done async in the background